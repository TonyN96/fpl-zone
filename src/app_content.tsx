import React from "react";
import { useQuery } from "react-query";
import { BrowserRouter as Router,Route, Routes } from "react-router-dom";
import { useMediaQuery } from "@mui/material";
import { getAllFixtures, getGameData } from "api/fpl_api_provider";
import {
  AccountPage,
  AnalysisPage,
  FixturesAndResultsPage,
  GameweekLivePage,
  LoginPage,
  MyTeamPage,
  RegisterPage,
  ResetPage
} from "pages";
import PrivateRoute from "private_route";
import { AppData } from "types";

import { Logout } from "components/authentication";
import { Notifier, Startup } from "components/layout";

export const AppDataContext = React.createContext<AppData | null>(null);

export default function AppContent(): JSX.Element {
  // Fetching game data which will be made available throughout the app via context provider
  const {
    data: gameData,
    isError: gameDataIsError,
    error: gameDataError,
    isLoading: gameDataIsLoading
  } = useQuery("game-data", getGameData);

  // Fetching fixture data which will be made available throughout the app via context provider
  const {
    data: fixtureData,
    isError: fixtureDataIsError,
    error: fixtureDataError,
    isLoading: fixtureDataIsLoading
  } = useQuery("all-fixtures", getAllFixtures);

  const isCompact = useMediaQuery("(max-width:1500px)");

  const isLoading = gameDataIsLoading || fixtureDataIsLoading;
  const isError = gameDataIsError || fixtureDataIsError;
  const appData = gameData && fixtureData ? { ...gameData, fixtures: fixtureData, isCompact } : null;

  if (isLoading) {
    // Display loading message if data is still being fetched
    return (
      <Startup>
        <Notifier />
      </Startup>
    );
  } else if (isError) {
    // Display error message if data fetch failed
    const error = gameDataError || fixtureDataError;
    const errorMessage = error instanceof Error ? `: ${error.message}` : ".";

    return (
      <Startup>
        <Notifier type='error' message={`An error has occured: ${errorMessage}`} />
      </Startup>
    );
  }

  // Otherwise, render the app
  return (
    <AppDataContext.Provider value={appData}>
      <Router>
        <Routes>
          <Route path='*' element={<LoginPage />} />
          <Route path='/' element={<LoginPage />} />
          <Route path='/login' element={<LoginPage />} />
          <Route path='/register' element={<RegisterPage />} />
          <Route path='/reset' element={<ResetPage />} />
          <Route path='/logout' element={<Logout />} />
          <Route path='/account' element={<PrivateRoute component={<AccountPage />} />} />
          <Route
            path='/gameweek-live'
            element={<PrivateRoute component={<GameweekLivePage />} />}
          />
          <Route path='/my-team' element={<PrivateRoute component={<MyTeamPage />} />} />
          <Route
            path='/fixtures-and-results'
            element={<PrivateRoute component={<FixturesAndResultsPage />} />}
          />
          <Route path='/analysis' element={<PrivateRoute component={<AnalysisPage />} />} />
        </Routes>
      </Router>
    </AppDataContext.Provider>
  );

}
