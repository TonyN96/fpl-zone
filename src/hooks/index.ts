export * from "./use_my_team_lineup";
export * from "./use_next_five_gameweek_ids";
export * from "./use_summary_stats";
