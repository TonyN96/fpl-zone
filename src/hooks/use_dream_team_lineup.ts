import { useContext } from "react";
import { AppDataContext } from "app_content";
import _ from "lodash";
import { AppData, Player } from "types";
import { Lineup } from "types/lineup";

export const useDreamTeamLineup = (): Lineup => {
  const { positions, players } = useContext(AppDataContext) as AppData;

  const squad = positions.map((pos): Player[] => _(players)
    .filter(["element_type", pos.id])
    .orderBy(["event_points"], ["desc"])
    .slice(0, pos.squad_select)
    .value()
  );

  const selected = squad.map((playersByPos, index) => playersByPos.splice(0, positions[index].squad_min_play));
  const bench: Player[] = [];

  const remainingPlayers = _(squad)
    .flatten()
    .orderBy(["event_points"], ["desc"])
    .value();

  remainingPlayers.forEach((player) => {
    const pos = player.element_type;
    const flatSelectedLen = selected.flat().length;
    const posSelectedLen = selected[pos - 1].length;
    const posSelectedMax = positions[pos - 1].squad_max_play;

    flatSelectedLen < 11 && posSelectedLen < posSelectedMax
      ? selected[pos - 1].push(player)
      : bench.push(player);
  });

  return { selected, bench };
};
