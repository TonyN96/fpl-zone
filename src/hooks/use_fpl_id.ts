import { useContext, useEffect, useState } from "react";
import { useAuthState } from "react-firebase-hooks/auth";
import { AuthContext } from "app_content";
import { auth, getUserFplTeamId } from "config";

export const useFplId = (): number | undefined => {
  const [user] = useAuthState(auth);
  const { fplId: savedFplId } = useContext(AuthContext);
  const [fplId, setFplId] = useState<number | undefined>();

  useEffect(() => {
    const getUserFplId = async (): Promise<void> => {
      const id = user ? await getUserFplTeamId(user.uid) : savedFplId;
      setFplId(id);
    };

    getUserFplId();
  }, [user, savedFplId]);

  return fplId;
};
