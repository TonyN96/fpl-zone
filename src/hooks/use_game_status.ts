import { useContext } from "react";
import { AppDataContext } from "app_content";
import { AppData } from "types";

interface GameStatusValues {
  seasonNotStarted: boolean;
  seasonFinished: boolean;
  gameUpdating: boolean;
}

/*
  This hook is used to check the status of the following:

  1. If the season has not yet started

  2. If the season has finished

  3. If the the most recent gameweek deadline was was within the last 1 hours
    - if true, FPL data is likely to be innacurate as the game is in an
    'updating' state for approx 1 hours after each gameweek deadline
*/
export const useGameStatus = (): GameStatusValues => {
  const { gameweeks } = useContext(AppDataContext) as AppData;

  let seasonNotStarted = false;
  let seasonFinished = false;
  let gameUpdating = false;

  const currentDateTime = new Date();
  const currentGameweek = gameweeks.find(gameweek => gameweek.is_current);
  const nextGameweek = gameweeks.find(gameweek => gameweek.is_next);

  if (!currentGameweek) {
    if (!nextGameweek) {
      seasonFinished = true;
    } else {
      seasonNotStarted = true;
    }
  }

  /*
    Manually checking if next gamewek deadline has passed as there
    can sometimes be a delay in update of gameweek 'is_next' status
  */
  const relevantGameweek = (!nextGameweek || new Date(nextGameweek.deadline_time) > currentDateTime)
    ? currentGameweek
    : nextGameweek;

  if (relevantGameweek) {
    const gameweekDeadline = new Date(relevantGameweek.deadline_time);
    const timeDifference = currentDateTime.getTime() - gameweekDeadline.getTime();

    // If current time is 1 hours (3600000 ms) since deadline, game will be in 'updating' state
    if (timeDifference < 3600000) {
      gameUpdating = true;
    }
  }

  return { seasonNotStarted, seasonFinished, gameUpdating };
};
