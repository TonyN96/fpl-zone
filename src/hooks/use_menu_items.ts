import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "app_content";
import { logout } from "config";

import { AuthModalContext, AuthModalView } from "components/layout";
import { MenuItemType } from "components/nav_drawer/menu_list";

interface MenuItems {
  nav: MenuItemType[];
  auth: MenuItemType[];
}

export const useMenuItems = (closeNavDrawer: () => void): MenuItems => {
  const { fplId, setFplId, user } = useContext(AuthContext);
  const navigate = useNavigate();
  const { setAuthModalView } = useContext(AuthModalContext);

  const onLogoutClick = (): void => {
    fplId ? setFplId() : logout();
  };

  const onAuthItemClick = (view: AuthModalView): void => {
    closeNavDrawer();
    setAuthModalView(view);
  };

  const nav: MenuItemType[] = [
    {
      id: "gw-live",
      label: "gameweek live",
      onClick: () => navigate("/gameweek-live"),
      subItems: [
        {
          id: "dream-team",
          label: "dream team",
          onClick: () => navigate("dream-team")
        },
        {
          id: "summary",
          label: "summary",
          onClick: () => navigate("summary")
        }
      ]
    },
    {
      id: "my-fpl",
      label: "my fpl",
      onClick: () => navigate("/my-fpl"),
      subItems: [
        {
          id: "my-team",
          label: "my team",
          onClick: () => navigate("my-team")
        },
        {
          id: "my-fpl-fdr",
          label: "fdr",
          onClick: () => navigate("my-fpl-fdr")
        }
      ]
    },
    {
      id: "fix-and-res",
      label: "fixtures & results",
      onClick: () => navigate("/fix-and-res"),
      subItems: [
        {
          id: "fix-and-res-fdr",
          label: "fdr",
          onClick: () => navigate("fix-and-res-fdr")
        },
        {
          id: "results",
          label: "results",
          onClick: () => navigate("results")
        }
      ]
    },
    {
      id: "analysis",
      label: "analysis",
      onClick: () => navigate("/analysis"),
      subItems: [
        {
          id: "comparison",
          label: "comparison",
          onClick: () => navigate("comparison")
        }
      ]
    }
  ];

  const auth: MenuItemType[] = [
    ...user || fplId
      ? [
        {
          id: "logout",
          label: "logout",
          onClick: onLogoutClick
        },
        ...user
          ? [
            {
              id: "account",
              label: "account",
              onClick: () => onAuthItemClick("account")
            }
          ]
          : []
      ]
      : [
        {
          id: "login",
          label: "login",
          onClick: () => onAuthItemClick("login")
        },
        {
          id: "register",
          label: "register",
          onClick: () => onAuthItemClick("register")
        }
      ]
  ];

  return { nav, auth };
};
