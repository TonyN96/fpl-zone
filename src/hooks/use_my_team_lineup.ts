import { useContext } from "react";
import { AppDataContext } from "app_content";
import { chunk } from "lodash";
import { AppData, Player, TeamPicks } from "types";
import { Lineup } from "types/lineup";

export const useMyTeamLineup = (teamPicks: TeamPicks): Lineup => {
  const { positions, players } = useContext(AppDataContext) as AppData;
  const [selectedPicks, benchPicks] = chunk(teamPicks.picks, 11);

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const getPlayerById = (id: number): Player => players.find((player) => player.id === id)!;

  const selected = positions.map((pos => {
    const picks = selectedPicks.filter(pick => getPlayerById(pick.element).element_type === pos.id);

    return picks.map((pick) => getPlayerById(pick.element));
  }));

  const bench = benchPicks.map((pick) => getPlayerById(pick.element));

  return { selected, bench };
};
