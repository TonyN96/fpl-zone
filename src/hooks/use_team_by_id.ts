import { useContext } from "react";
import { AppDataContext } from "app_content";
import { AppData, Team } from "types";

export const useTeamById = (id: number): Team => {
  const { teams } = useContext(AppDataContext) as AppData;

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return teams.find((team) => team.id === id)!;
};
