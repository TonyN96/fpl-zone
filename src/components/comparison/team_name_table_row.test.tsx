import React from "react";
import { render, screen } from "@testing-library/react";
import { AppDataContext } from "app_content";
import { useTeamById } from "hooks/use_team_by_id";
import { mockAppData, mockPlayers } from "test/test_data";

import "@testing-library/jest-dom/extend-expect";

import { MAX_PLAYER_COUNT, TeamNameTableRow } from ".";

describe("Team name table row tests", () => {
  let mockSelectedPlayers = mockPlayers.splice(0, 3);

  const createComponent = (): JSX.Element => {
    return (
      <AppDataContext.Provider value={mockAppData}>
        <TeamNameTableRow
          selectedPlayers={mockSelectedPlayers}
        />
      </AppDataContext.Provider>
    );
  };

  it("Player's team name displayed correctly", () => {
    render(createComponent());

    mockSelectedPlayers.forEach((player) => {
      const row = screen.getByTestId(`team-name-row-${player.id}`);
      const team = useTeamById(player.team);
      expect(row).toHaveTextContent(team.name);
    });
  });

  describe("Empty table cell", () => {
    it(`displayed when less than ${MAX_PLAYER_COUNT} players are selected`, () => {
      render(createComponent());

      expect(screen.getByTestId("empty-table-cell")).toBeInTheDocument();
    });

    it(`not displayed when ${MAX_PLAYER_COUNT} players are selected`, () => {
      mockSelectedPlayers = mockPlayers.slice(0, MAX_PLAYER_COUNT);

      render(createComponent());

      expect(screen.queryByTestId("empty-table-cell")).toBeNull();
    });
  });
});
