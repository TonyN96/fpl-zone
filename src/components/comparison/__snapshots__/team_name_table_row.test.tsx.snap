// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team name table row tests Snapshot test 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Team
  </p>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Aston Villa
  </p>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Crystal Palace
  </p>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Arsenal
  </p>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Leicester
  </p>
</DocumentFragment>
`;
